### ISSUE TITLE
CsvHelper.Configuration.IWriterConfiguration.TypeConverterOptionsCache no longer accessible in v20

### ISSUE THREAD

**Issue by shandysawyer** [JoshClose/CsvHelper#1645]
**Describe the bug**
CsvHelper.Configuration.IWriterConfiguration.TypeConverterOptionsCache is no longer an accessible member in v20.0.0. Not sure if this was done intentionally? If intentional, what is the prescribed method now for something like so?

```csharp
var csv = new CsvWriter(sw, CultureInfo.InvariantCulture);
var options = new TypeConverterOptions { Formats = new[] { "MM/dd/yyyy" } };
csv.Configuration.TypeConverterOptionsCache.AddOptions<DateTime>(options);
csv.Configuration.TypeConverterOptionsCache.AddOptions<DateTime?>(options);
```

**Comment by JoshClose** [comment]
It's in the context now.

**Comment by JoshClose** [comment]
https://joshclose.github.io/CsvHelper/change-log/

**Comment by shandysawyer** [comment]
Ahh I see, so now it goes like this:

```csharp
csv.Context.get_TypeConverterOptionsCache().AddOptions<DateTime>(options);
```

Surprised to see all the "get_" prefixes now. Thank you for the very quick follow-up.

Closing the issue.

**Comment by JoshClose** [comment]
Oh there shouldn't be prefixes... I'll take a look in the morning when I get to my computer.

**Comment by JoshClose** [comment]
I'm not seeing `get_TypeConverterOptionsCache`. I downloaded it on LINQPad and it looks normal. Where are you seeing this?

**Comment by shandysawyer** [comment]
Strange, I downloaded v20 off of NuGet for a .NET Core Application:
![csvhelper1](https://user-images.githubusercontent.com/867743/104933144-8528f900-5976-11eb-9012-97ee263e6241.png)
![csvhelper2](https://user-images.githubusercontent.com/867743/104933164-8fe38e00-5976-11eb-8f82-d02b94d6f334.png)

**Comment by JoshClose** [comment]
I'll take a look tomorrow. I'll re-open so I don't forget about this.

**Comment by JoshClose** [comment]
There was an issue with the netstandard version being 1.0.0 for some reason so maybe that was causing this. Can you download the latest version and see if it's resolved?

**Comment by shandysawyer** [comment]
Unfortunately the prefix is still there

![csvhelper1](https://user-images.githubusercontent.com/867743/105073219-c55ea900-5a54-11eb-9243-f405ff3eab4a.png)
![csvhelper2](https://user-images.githubusercontent.com/867743/105073226-c7286c80-5a54-11eb-92cc-c1d38203e06e.png)

**Comment by JoshClose** [comment]
That's just in intellisense?

**Comment by shandysawyer** [comment]
Here's the class definition I'm seeing:

![csvhelper3](https://user-images.githubusercontent.com/867743/105077300-53895e00-5a5a-11eb-9f6e-7b3a0e4b4b5d.png)

**Comment by shandysawyer** [comment]
Visual Studio Professional 2019 version 16.4.4 
.NET Framework Version 4.8.03761

Steps to replicate:
1. Create a new .NET Core 3.1 Console App project
2. Right click Dependencies -> Manage NuGet Packages
3. Search "CsvHelper" -> Click Install (latest version automatically selected)
4. Navigate to class definition of CsvContext

**Comment by JoshClose** [comment]
I get this:
![image](https://user-images.githubusercontent.com/114971/105200168-2940aa80-5b05-11eb-9bc6-2b9a5bbff987.png)

My Visual Studio is 16.8.4. Could you try updating Visual Studio? I'm not able to reproduce this in any environment.

**Comment by shandysawyer** [comment]
Working on getting it updated, no admin rights to do so myself.

**Comment by JoshClose** [comment]
I'm sorry to hear that. I've been there.

**Comment by HaMa84** [comment]
There is another issue. Using custom type converters for custom types used in anonymous objects does not work anymore. The reason is, that auto mapping creates a new instance of CsvContext that contains only the default type converters. The issue started with version 20.

**Comment by JoshClose** [comment]
Can you give an example of how you're using it? There is a new overload `AutoMap(CsvContext context)` that might be what you're looking for. If not, I'll see what I can do based on your example.

**Comment by HaMa84** [comment]
The console output of the following example is

TheValue
string

but should be

str
string

Worked as expected up to version 19. I don't know how to solve this.

using CsvHelper;
using CsvHelper.Configuration;
using CsvHelper.TypeConversion;
using System;
using System.Globalization;
using System.IO;
using System.Text;
 
namespace ConsoleApp
{
    public class TheString
    {
        public string TheValue { get; set; }
    }
 
    public class TheStringConverter : ITypeConverter
    {
        public object ConvertFromString(string text, IReaderRow row, MemberMapData memberMapData)
            => throw new NotImplementedException();
 
        public string ConvertToString(object value, IWriterRow row, MemberMapData memberMapData)
            => (value as TheString)?.ToString() ?? "";
    }
 
    class Program
    {
        static void Main()
        {
            var cultureInfo = CultureInfo.CurrentCulture;
            var sb = new StringBuilder();
            using (var writer = new StringWriter(sb))
            using (var csv = new CsvWriter(writer, new CsvConfiguration(cultureInfo)))
            {
                csv.Context.TypeConverterCache.AddConverter<TheString>(new TheStringConverter());
                csv.WriteRecords(new[] { new { str = new TheString { TheValue = "string" } } });
            }
            Console.Write(sb.ToString());
            Console.ReadLine();
        }
    }
}

**Comment by JoshClose** [comment]
I see the where the issue is happening. I'm going to create a new issue for this since it's unrelated to the OP's issue.

**Comment by LeorGreenberger** [comment]
I ran into this issue today as well and I may have some insight into what I think is going on. When you build this with MSBuild v16.6 or older, I get an error:
```
ExportService.cs(47,29): error CS1545: Property, indexer, or event 
'CsvContext.TypeConverterOptionsCache' is not supported by the language; try directly calling 
accessor methods 'CsvContext.get_TypeConverterOptionsCache()' or 
'CsvContext.set_TypeConverterOptionsCache(?)' [/src/TMP.MetricsService.AppService
/TMP.MetricsService.AppService.csproj]
```

When using 16.7 or later of MSBuild, there is no such error. 
I ran into this issue today when trying to build it on our build server. I ended up noticing that on the build server, I was getting the following in the build output:

```
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
```

while locally I was getting:

```
Microsoft (R) Build Engine version 16.7.2+b60ddb6f4 for .NET
```

**Comment by JoshClose** [comment]
Thanks! I figured that would be the issue. Updating Visual Studio should fix that then, right?

**Comment by LeorGreenberger** [comment]
> 
> 
> Thanks! I figured that would be the issue. Updating Visual Studio should fix that then, right?

Yes. In my case, I am building it in docker and in the dockerfile, I have 
```
FROM mcr.microsoft.com/dotnet/core/sdk:3.1-buster AS build-env
```
And I confirmed that this is using the latest .net core 3.1 SDK, but it shows MSBuild 16.6 in the build output. When I changed that line to

```
FROM mcr.microsoft.com/dotnet/sdk:3.1 AS build-env
```

It shows 16.7 in the build output. It's very strange.

**Comment by shandysawyer** [comment]
Unfortunately I am STILL waiting for admins to update Visual Studio on the machine I use for development. However it sounds like @LeorGreenberger ran into the same issue and saw the issue go away when using the latest build engine. Also it appears others aren't running into this issue as well otherwise I would expect more to be chiming in. I'm going to close this unless @JoshClose wants to keep it open to investigate @LeorGreenberger issue as well.

**Comment by JoshClose** [comment]
Let's wait and see if a Visual Studio update fixes it for you.

**Comment by JoshClose** [comment]
@HaMa84 The issue you were seeing is fixed in version 23.3.1 in NuGet. https://www.nuget.org/packages/CsvHelper/21.3.1

**Comment by shandysawyer** [comment]
Confirmed, upgrading to 16.8.4 resolved that issue of the "get_" prefix. 

I'm having a hard time understanding the new work flow, and unfortunately the documentation is still under construction with TypeConversion. Is the idea to do something like this now?

```csharp
var csvConfig = new CsvConfiguration(CultureInfo.InvariantCulture) { HasHeaderRecord = true, Delimiter = "|" };
var csvContext = new CsvContext(csvConfig);

// remove timestamps from any DateTime fields
var dateFormatOptions = new TypeConverterOptions { Formats = new[] { "MM/dd/yyyy" } };
csvContext.TypeConverterOptionsCache.AddOptions<DateTime>(dateFormatOptions);
csvContext.TypeConverterOptionsCache.AddOptions<DateTime?>(dateFormatOptions);

// I'm guessing csvConfig doesn't know anything about csvContext when doing this?
var csvWriterOld = new CsvWriter(new StreamWriter(".\file.txt"), csvConfig);

// So is the idea to do this instead to get the proper CsvWriter with the type converter options enabled?
IWriter csvWriterNew = csvContext.Writer;
```

**Comment by JoshClose** [comment]
The config is read only and you have to create it ahead of time and pass it into the reader/writer. The context is available on the reader/writer and isn't immutable, so you can do `csvWriter.Context.TypeConverterOptionsCache...`.

The context is created in the constructor of the reader/writer and will have the configuration you pass in.

**Comment by JoshClose** [comment]
Here is an example.

```cs
void Main()
{
	var records = new List<Foo>
	{
		new Foo { Id = 1, Name = "one" },
	};
	var config = new CsvConfiguration(CultureInfo.InvariantCulture)
	{
		ShouldQuote = (_, _, _) => true,
	};
	using (var writer = new StringWriter())
	using (var csv = new CsvWriter(writer, config))
	{
		csv.Context.RegisterClassMap<FooMap>();
		csv.WriteRecords(records);
		writer.Dump();
	}
}

public class Foo
{
	public int Id { get; set; }
	public string Name { get; set; }
}

public class FooMap : ClassMap<Foo>
{
	public FooMap()
	{
		Map(m => m.Id);
		Map(m => m.Name);
	}
}
```

**Comment by Josep313** [comment]
Could this potentially be re-opened?

I am using CSVHelper 22.1.2 with VS 2017. Our MSBuild is 15.x which seems to create the prefixing issue with CSVConfiguration

**Comment by JoshClose** [comment]
Can you see if version 23.0.0 fixes that? I removed the `init` properties. I don't know if it is `init` or `record` that is doing that.

**Comment by zacrumford** [comment]
I am also hitting this. I upgraded to v25.0.0 built locally on my vs2019 machine and it worked a-ok. On the official build machines I see this issue.

**Comment by JoshClose** [comment]
This is fixed in 26.0.0 on NuGet. https://www.nuget.org/packages/CsvHelper/26.0.0

I removed all `init` properties.

**Comment by oryz** [comment]
Sorry to revive this, but encountered this issue on an upgrade from v16 to v27.

I've updated our code to use 
`csv.Context.TypeConverterOptionsCache.GetOptions<string>()...`

instead of - 
`csv.Configuration.TypeConverterOptionsCache.GetOptions<string>()...`

This works in most places, but not when we're trying to share this configuration between a few different csvReader instances e.g.:

`
            // code

            List<Member> memberRecords;
            List<Policy> policyRecords;

            var config = new CsvHelper.Configuration.CsvConfiguration(CultureInfo.InvariantCulture)
            {
                HasHeaderRecord = true,
                HeaderValidated = null,
                MissingFieldFound = null,
            };
            config.TypeConverterOptionsCache.GetOptions<string>().NullValues.Add("null");
            config.TypeConverterOptionsCache.GetOptions<bool>().NullValues.Add("");
            // Read members
            using (var reader = new StreamReader(csvFilePath))
            {
                using (var csv = new CsvReader(reader, config))
                {
                    memberRecords = csv.GetRecords<Member>().Where(x => x.Import == true)
                        .ToList();
                }
            }

            // read policies
            using (var reader = new StreamReader(csvFilePath))
            {
                using (var csv = new CsvReader(reader, config))
                {
                    csv.Context.RegisterClassMap<PolicyMap>();
                    policyRecords = csv.GetRecords<Policy>().Where(x => x.Import == true).ToList();
                }
            }

Is there a way to do this in the newer implementation or would I have to specify `TypeConverterOptionsCache.GetOptions<>` inside each csvReader block?
